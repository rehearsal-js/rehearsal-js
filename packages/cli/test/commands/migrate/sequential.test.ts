import { resolve } from 'node:path';
import { readdirSync } from 'node:fs';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { readJSONSync } from 'fs-extra/esm';
import { createLogger, format, transports } from 'winston';
import {
  initTask,
  depInstallTask,
  tsConfigTask,
  lintConfigTask,
  sequentialTask,
} from '../../../src/commands/migrate/tasks/index.js';

import {
  prepareTmpDir,
  listrTaskRunner,
  createOutputStream,
  createMigrateOptions,
} from '../../test-helpers/index.js';

const logger = createLogger({
  transports: [new transports.Console({ format: format.cli() })],
});

describe('Task: sequential', async () => {
  let basePath = '';
  let output = '';

  let outputStream = createOutputStream();

  vi.spyOn(console, 'info').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'log').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });

  beforeEach(() => {
    output = '';
    basePath = prepareTmpDir('sequential');
    outputStream = createOutputStream();
  });

  afterEach(() => {
    output = '';
    vi.clearAllMocks();
    outputStream.destroy();
  });

  test('sequential run regen on the existing report, and run migrate on current base path and entrypoint', async () => {
    const options = createMigrateOptions(basePath, { entrypoint: 'index.ts' });
    const previousRuns = {
      previousFixedCount: 1,
      paths: [{ basePath, entrypoint: 'depends-on-foo.ts' }],
    };
    const tasks = [
      await initTask(options),
      await depInstallTask(options),
      await tsConfigTask(options),
      await lintConfigTask(options),
      await sequentialTask(options, logger, previousRuns),
    ];

    await listrTaskRunner(tasks);
    expect(output).toMatchSnapshot();

    const fileList = readdirSync(basePath);
    expect(fileList).toContain('depends-on-foo.ts');
    expect(fileList).toContain('foo.ts');
    expect(fileList).toContain('index.ts');

    const report = readJSONSync(resolve(basePath, '.rehearsal', 'migrate-report.json'));
    const { summary, fixedItemCount, items } = report;
    expect(summary.length).toBe(2);
    expect(summary[0].basePath).toEqual(summary[1].basePath);
    expect(summary[0].entrypoint).toBe('depends-on-foo.ts');
    expect(summary[1].entrypoint).toBe('index.ts');
    expect(fixedItemCount).toBe(2);
    expect(items.length).toBe(2);
  });
});
