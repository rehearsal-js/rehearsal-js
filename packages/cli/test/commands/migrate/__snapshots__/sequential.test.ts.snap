// Vitest Snapshot v1

exports[`Task: sequential > sequential run regen on the existing report, and run migrate on current base path and entrypoint 1`] = `
"[STARTED] Initialize
[DATA] Running migration on sequential
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Regenerating past migrate report(s), generating current migrate report and merging all reports
[DATA] processing file: /foo.ts
[DATA] processing file: /index.ts
[DATA] processing file: /depends-on-foo.ts
[TITLE] Rescanning previously migrated files based on existing report:
[TITLE] 
[TITLE]   3 ts files have been scanned
[TITLE] 
[TITLE]   4 errors caught by rehearsal, and will be merged into the migration report
[TITLE] 
[TITLE]     
[DATA] processing file: /foo.ts
[DATA] processing file: /index.ts
[TITLE] Migration Complete
[TITLE] 
[TITLE] 
[TITLE]   2 JS files converted to TS
[TITLE] 
[TITLE]   6 errors caught by rehearsal
[TITLE] 
[TITLE]   4 have been fixed by rehearsal
[TITLE] 
[TITLE]   2 errors need to be fixed manually
[TITLE] 
[TITLE]     -- 1 ts errors, marked by @ts-expect-error @rehearsal TODO
[TITLE] 
[TITLE]     -- 1 eslint errors, with details in the report
[TITLE] 
[SUCCESS] Migration Complete
[SUCCESS] 
[SUCCESS] 
[SUCCESS]   2 JS files converted to TS
[SUCCESS] 
[SUCCESS]   6 errors caught by rehearsal
[SUCCESS] 
[SUCCESS]   4 have been fixed by rehearsal
[SUCCESS] 
[SUCCESS]   2 errors need to be fixed manually
[SUCCESS] 
[SUCCESS]     -- 1 ts errors, marked by @ts-expect-error @rehearsal TODO
[SUCCESS] 
[SUCCESS]     -- 1 eslint errors, with details in the report
[SUCCESS] 
"
`;
